main/src/ca/uwaterloo/flix/api/Bootstrap.scala

```diff
@@ -825,4 +832,21 @@ class Bootstrap(val projectPath: Path, apiKey: Option[String]) {
       Validation.Success(true)
     }
   }
+
+  def effectLock(path: Path, flix: Flix): Validation[Unit, BootstrapError] = {
+    check(flix) match {
+      case Validation.Failure(errors) => Validation.Failure(errors)
+      case Validation.Success(root) =>
+        val outputStream = Files.newOutputStream(getEffectLockFile(path))
+        val signatures = mkEffectLockSignatures(root).mkString("\n") // TODO: make JSON
+        Validation.Success(outputStream.write(signatures.getBytes))
+    }
+  }
+
+  def mkEffectLockSignatures(root: TypedAst.Root): List[String] = {
```

Comment:
I suggest a separate compiler phase. But perhaps even better, we should extend the Safety phase.

Reply:
Yes agree. Was exploring code here that I would refactor before marking the pr green

Reply:
Yeah, lets get some code in and then we can iterate.